src/cluster/placement/algo/mirrored_test.go

```diff
@@ -646,37 +429,318 @@ func TestMirrorRemoveAndRevertAfterCutover(t *testing.T) {
-		SetID("i3").
-		SetIsolationGroup("r3").
-		SetEndpoint("endpoint3").
-		SetShardSetID(2).
-		SetWeight(2)
-	i4 := placement.NewInstance().
-		SetID("i4").
-		SetIsolationGroup("r4").
-		SetEndpoint("endpoint4").
-		SetShardSetID(2).
-		SetWeight(2)
-	i5 := placement.NewInstance().
-		SetID("i5").
-		SetIsolationGroup("r5").
-		SetEndpoint("endpoint5").
-		SetShardSetID(2).
-		SetWeight(2)
+	i1 := newTestInstance("i1").SetShardSetID(1).SetWeight(1)
+	i2 := newTestInstance("i2").SetShardSetID(1).SetWeight(1)
+	i3 := newTestInstance("i3").SetShardSetID(2).SetWeight(2)
+	i4 := newTestInstance("i4").SetShardSetID(2).SetWeight(2)
+	i5 := newTestInstance("i5").SetShardSetID(2).SetWeight(2)
+	instances := []placement.Instance{i1, i2, i3, i4}
+
+	numShards := 100
+	ids := make([]uint32, numShards)
+	for i := 0; i < len(ids); i++ {
+		ids[i] = uint32(i)
+	}
+
+	now := time.Now()
+	nowNanos := now.UnixNano()
+	shardCutoverTime := now.Add(time.Hour).UnixNano()
+	a := NewAlgorithm(placement.NewOptions().
+		SetIsMirrored(true).
+		SetShardCutoverNanosFn(func() int64 { return shardCutoverTime }).
+		SetShardCutoffNanosFn(func() int64 { return shardCutoverTime }))
+	p, err := a.InitialPlacement(instances, ids, 2)
+	assert.NoError(t, err)
+	assert.NoError(t, placement.Validate(p))
+	assert.Equal(t, uint32(2), p.MaxShardSetID())
+
+	p1, err := a.ReplaceInstances(p, []string{"i4"}, []placement.Instance{i5})
+	assert.NoError(t, err)
+	assert.NoError(t, placement.Validate(p1))
+	assert.Equal(t, uint32(2), p1.MaxShardSetID())
+	assert.True(t, allLeaving(p1, []placement.Instance{i4}, nowNanos))
+	assert.True(t, allInitializing(p1, []string{"i5"}, nowNanos))
+
+	p2, err := a.ReplaceInstances(p1, []string{"i5"}, []placement.Instance{i4})
+	assert.NoError(t, err)
+	assert.NoError(t, placement.Validate(p2))
+	assert.Equal(t, uint32(2), p2.MaxShardSetID())
+	i4, ok := p2.Instance("i4")
+	assert.True(t, ok)
+	assert.Equal(t, i4.Shards().NumShards(), i4.Shards().NumShardsForState(shard.Available))
+	_, ok = p2.Instance("i5")
+	assert.False(t, ok)
+}
+
+func TestMirrorReplaceAndRevertAfterCutover(t *testing.T) {
+	i1 := newTestInstance("i1").SetShardSetID(1).SetWeight(1)
+	i2 := newTestInstance("i2").SetShardSetID(1).SetWeight(1)
+	i3 := newTestInstance("i3").SetShardSetID(2).SetWeight(2)
+	i4 := newTestInstance("i4").SetShardSetID(2).SetWeight(2)
+	i5 := newTestInstance("i5").SetShardSetID(2).SetWeight(2)
+	instances := []placement.Instance{i1, i2, i3, i4}
+
+	numShards := 100
+	ids := make([]uint32, numShards)
+	for i := 0; i < len(ids); i++ {
+		ids[i] = uint32(i)
+	}
+
+	now := time.Now()
+	nowNanos := now.UnixNano()
+	shardCutoverTime := now.Add(-time.Hour).UnixNano()
+	a := NewAlgorithm(placement.NewOptions().
+		SetIsMirrored(true).
+		SetShardCutoverNanosFn(func() int64 { return shardCutoverTime }).
+		SetShardCutoffNanosFn(func() int64 { return shardCutoverTime }))
+	p, err := a.InitialPlacement(instances, ids, 2)
+	assert.NoError(t, err)
```

Comment:
Nit (here and elsewhere): pretty much always worth doing `require.NoError`--otherwise tests tend to panic when they barrel ahead with nils.

Reply:
Agree, I missed this one and limited it only to newly added tests to avoid making this diff bloated.

Reply:
Ah yeah sure, this looked new to me but it may not be.